### ISSUE TITLE
识别问题

### ISSUE THREAD

**Issue by Mnanazi** [yeyupiaoling/VoiceprintRecognition-Pytorch#6]
输入一个为注册用户的语音，但是匹配到了声纹库中的一个已注册用户，而且还给出相似度0.998...(情况如下图)。

说明：addio_db中的声纹音频内容都是以注册用户名称开头的，例如 123.wav 表示 注册时录音 “123456” ；机器人识别.wav 表示 注册时录音 "机器人识别"
前提1：只用了2000条音频数据训练+测试，而且更改了训练模型的几个参数，learning_rate=0.1, batch_size=8,num_epoch=10
前提2：我也出现了issue中“关于训练”中   “loss: 17.497463, accuracy: 0.000000”,loss正常但是accuracy全部为0的情况  
疑问：请问是什么原因导致这种声纹匹配错误的情况下（明明是一个从未在注册库中的陌生说话人录音，但却匹配到了声纹库中的一个注册用户，还给出了0.99的相似度评价），还给出了0.99的相似度评价？

![image](https://user-images.githubusercontent.com/75676487/136643325-527530c4-13e3-4c0e-85d1-d8757b77e70a.png)

**Comment by Mnanazi** [comment]
> 你自己训练的模型吗？你用的数据太少了。

因为将mp3转换为wav太花时间了，昨天转换8万条花了6个多小时。
能不能分享下全部转换为wav的音频数据？？

**Comment by Mnanazi** [comment]
> > 你自己训练的模型吗？你用的数据太少了。
> 
> 因为将mp3转换为wav太花时间了，昨天转换8万条花了6个多小时。 能不能分享下全部转换为wav的音频数据？？
对，是自己训练的模型

**Comment by Mnanazi** [comment]
> 转换后数据很大，分享不了。

好吧。我就先用8万条数据训练下模型

**Comment by yeyupiaoling** [comment]
要就用全部数据，这个数据太少了。

**Comment by Mnanazi** [comment]
> 要就用全部数据，这个数据太少了。

类别数多少怎么看，分类类别数的含义：是语料中总记得说话人人数吗？你在trian.py中的分类类别数设置为3242，如果分类类别是确实指的是说话人总数，那我根据train_list.txt中的数据，是不是应该把分类类别数设置为245（见图2）
![image](https://user-images.githubusercontent.com/75676487/136650444-21fc5bc8-6396-4ada-8014-8ed0856c362f.png)
图1
![image](https://user-images.githubusercontent.com/75676487/136650788-ff92706f-362c-4f04-9063-54c12eb92966.png)
图2

**Comment by Mnanazi** [comment]
> 是的，分类就是分这个

> 因为这个数据集没有json文件，所以我用的您另一个项目的这个文件生成的https://github.com/yeyupiaoling/VoiceprintRecognition-Tensorflow/blob/master/create_data.py 生成的train_list.txt，共42570条数据，应该也可以在这个项目训练吧？
> 
> 

我重新设置类别数后，遇到了trian时也遇到了accuracy=0的问题，我的类别数为246，train_list.txt最大数为245. 请问您是怎么解决accuracy=0的问题的？
![image](https://user-images.githubusercontent.com/75676487/136662622-dd2c7025-30e0-4514-8be7-07a3b9edbc81.png)
![image](https://user-images.githubusercontent.com/75676487/136662633-f55f8c30-564e-46e6-b4e4-2fb0bf16a4c3.png)

**Comment by Mnanazi** [comment]
> 你才刚开始训练，先训练20个epoch你再看看吧

正在训练，给出的eta为大于1天，绝望。笔记本风扇呼呼直响

**Comment by yeyupiaoling** [comment]
用笔记本训练？你没有GPU服务器吗？

**Comment by Mnanazi** [comment]
> 用笔记本训练？你没有GPU服务器吗？
是用笔记本训练的。没用过gpu服务器，请问用gpu服务器训练很快吗，一般多久？

以下是废话：
另外，训练时期间遇到一个bug，在第一epoch快要训练完的时候遇到这个错误ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1,512,1,1])
我看了下，训练时batch是100个样本来训练的，训练到最后不足100个样本，就报出上面的错误。
后来我在train_loader和test_loader中的DataLoader函数都添加了drop_last=True参数，之后训练完第一个epoch没有报错。
说这个报错问题·，主要是为了后面的人·遇到了该情况好解决。

**Comment by Mnanazi** [comment]
> 录音的内容是没有关系的，你其他的音频是自己拷贝进去的？还是多个人录音的？

测试的几个音频都是我一个人直接录音的，不是拷贝的，

**Comment by yeyupiaoling** [comment]
音频库里面的音频是你从其他地方裁剪过来的？有没有去掉镜片段，采样率是不是16KHz

**Comment by yeyupiaoling** [comment]
....怎么能都是你录的呢？应该是每个音频对应不同的人。这是声纹识别，不是语音识别

**Comment by Mnanazi** [comment]
> > 说错了，是静音片段。音频库里面是存放不同人录音的音频文件。我看你的意思是全部都是你自己录入进去的？
> 
> 对的，全部是我自己录的音。

是需要不同的说话人录音吗？ 好像也应该是这样，毕竟这个模型是对不同说话人进行分类的。 请问我这样理解有错吗？

**Comment by yeyupiaoling** [comment]
你自己摸索吧

**Comment by Mnanazi** [comment]
> > > 说错了，是静音片段。音频库里面是存放不同人录音的音频文件。我看你的意思是全部都是你自己录入进去的？
> > 
> > 
> > 对的，全部是我自己录的音。
> 
> 是需要不同的说话人录音吗？ 好像也应该是这样，毕竟这个模型是对不同说话人进行分类的。 请问我这样理解有错吗？



> 你自己摸索吧
兄弟这次测试成功了。测试了两个人，分别说不同的话，然后各自重复之前的话，识别成功。同时说话人A重复说话人B的话，不会被模型判断是同一个人。
但是，刚刚把infer_recognitin程序停下来，再程序启动，说话人A仍然重复之前的话，就会碰到有时识别的出说话人有时识别不出说话人的情况(试了重复启动程序3次，只有一次识别出来了)。

**Comment by Mnanazi** [comment]
测试发现，同一个说话人说完全不同的话，不会被判定是同一个说话人，这是模型效果不好的表现吧？起码给个低分如0.3的匹配度吧

> 注意不要有静音。音频库里面的音频也不要有静音
没静音，就是正常说话的语速

**Comment by Mnanazi** [comment]
> 分数不能设置太低。 要不你试一下其他的项目。

你发布的那个训练模型是使用全部数据集110万条语料训练的吗
